src/protocol/malicious.rs

```diff
@@ -115,40 +127,59 @@ impl<F: Field> SecurityValidatorAccumulator<F> {
 }
 
 #[allow(dead_code)]
-pub struct SecurityValidator<F: Field> {
+pub struct MaliciousValidator<'a, F: Field> {
     r_share: Replicated<F>,
     u_and_w: Arc<Mutex<AccumulatorState<F>>>,
+    protocol_ctx: MaliciousContext<'a, F>,
+    propagate_uw_ctx: SemiHonestContext<'a, F>,
+    reveal_r_ctx: SemiHonestContext<'a, F>,
+    check_zero_ctx: SemiHonestContext<'a, F>,
 }
 
-impl<F: Field> SecurityValidator<F> {
+impl<'a, F: Field> MaliciousValidator<'a, F> {
     #[must_use]
     #[allow(clippy::needless_pass_by_value)]
-    pub fn new(ctx: SemiHonestContext<'_, F>) -> SecurityValidator<F> {
+    pub fn new(ctx: SemiHonestContext<'a, F>) -> MaliciousValidator<F> {
+        // Use the current step in the context for initialization.
         let prss = ctx.prss();
-
         let r_share = prss.generate_replicated(RECORD_0);
-
         let state = AccumulatorState {
             u: prss.zero(RECORD_1),
             w: prss.zero(RECORD_2),
         };
 
-        SecurityValidator {
+        let u_and_w = Arc::new(Mutex::new(state));
+        let accumulator = MaliciousValidatorAccumulator {
+            inner: Arc::downgrade(&u_and_w),
+        };
+        let propagate_uw_ctx = ctx.narrow(&Step::PropagateUW);
+        let reveal_r_ctx = ctx.narrow(&Step::RevealR);
+        let check_zero_ctx = ctx.narrow(&Step::CheckZero);
+        let protocol_ctx = ctx.upgrade(
+            &Step::MaliciousProtocol,
+            &Step::UpgradeInput,
+            accumulator,
+            r_share.clone(),
+        );
+        MaliciousValidator {
             r_share,
-            u_and_w: Arc::new(Mutex::new(state)),
-        }
-    }
-
-    pub fn accumulator(&self) -> SecurityValidatorAccumulator<F> {
-        SecurityValidatorAccumulator {
-            inner: Arc::downgrade(&self.u_and_w),
+            u_and_w,
+            protocol_ctx,
+            propagate_uw_ctx,
+            reveal_r_ctx,
+            check_zero_ctx,
```

Comment:
Why do it this way?

It seems like you're taking just one context, and narrowing it 3 times, only to save all 3 of those values until validate is called.

At the time validate is called however, you still call `narrow`, specifically you call
`check_zero_ctx.narrow(&Step::CheckZero)`

Which makes me wonder: why not just pass in one context, save just one context, and in the call to validate just call narrow a few times?

Reply:
I shouldn't have been narrowing in validate, but I've taken another go at this to drop the overheads a little.

The problem here is that I made `ctx.upgrade()` consume `ctx`, so this needs to reserve some narrowed contexts for later.  I think that is the right model for this, even if `upgrade()` is only an internal thing.